coord/coordinator.go

```diff
@@ -0,0 +1,285 @@
+package coord
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/benbjohnson/clock"
+
+	"github.com/plprobelab/go-kademlia/kad"
+	"github.com/plprobelab/go-kademlia/network/address"
+	"github.com/plprobelab/go-kademlia/network/endpoint"
+	"github.com/plprobelab/go-kademlia/query"
+	"github.com/plprobelab/go-kademlia/util"
+)
+
+// A Coordinator coordinates the state machines that comprise a Kademlia DHT
+type Coordinator[K kad.Key[K], A kad.Address[A]] struct {
+	clk clock.Clock
+
+	qp *query.QueryPool[K, A]
+
+	// rt is the routing table used to look up nodes by distance
+	rt kad.RoutingTable[K]
+
+	// ep is the message endpoint used to send requests
+	ep endpoint.Endpoint[K, A]
+
+	peerstoreTTL time.Duration
```

Comment:
what do you think about letting `Coordinator` hold on to a `CoordinatorConfig` struct. That way we could remove the `clk` and `pperstoreTTL` fields from here.

Reply:
Maybe. It's just a stylistic choice. It would mean that the fields would always be referred to as `c.cfg.clk` rather than `c.clk`